### ISSUE TITLE
Some issues about training and validation processes

### ISSUE THREAD

**Issue by BruceWayne233** [HuguesTHOMAS/KPConv-PyTorch#140]
Hi Mr Thomas

First of all, than you for your amazing work，I used your work to separate the ground from the non-ground in the forest region，and  I‘ve got a mIou 80.32.

I have some issues about training and test processes.
I noticed that when reading the dataset, it will put all the rooms together，according to my understanding, the training is carried out on the whole, through in _ radius this parameter to select a region on the whole, this region is a ball. Then the “potential value“ is used to ensure that all regions can be calculated. The epoch _ steps parameter is used to control the number of balls selected in an epoch. If epoch _ steps is 500, it is equivalent to selecting 500 balls in an epoch on this whole.

With regard to the test process, it adds a voting mechanism to select the best result by traversing the entire test area many times.Whether the rest part of test process is similar to the training process ?

Is my understanding correct ? If there is an error please correct.

Finally, I ' m sorry for my bad English, I hope you can understand my meaning, I hope I can get your systematic reply about how the training and testing process works,

Thanks

**Comment by BruceWayne233** [comment]
by the way,where can I control the epochs of test?

**Comment by HuguesTHOMAS** [comment]
Hi @BruceWayne233,

Thanks for your interest in my code. I am happy to hear that it performed well on forest ground detection.

> I noticed that when reading the dataset, it will put all the rooms together, according to my understanding, the training is carried out on the whole, through in _ radius this parameter to select a region on the whole, this region is a ball. Then the “potential value“ is used to ensure that all regions can be calculated. The epoch _ steps parameter is used to control the number of balls selected in an epoch. If epoch _ steps is 500, it is equivalent to selecting 500 balls in an epoch on this whole.

Your understanding is correct except for one small detail. If `epoch_steps = 500`, it is equivalent to selecting 500 batches, which means approximatively `500 x batch_num` balls in an epoch.

> With regard to the test process, it adds a voting mechanism to select the best result by traversing the entire test area many times.Whether the rest part of test process is similar to the training process ?

Yes the test process follows the same strategy, using potential values. the number of spheres picked in the test process is controlled by `validation_size` instead of `epoch_steps`.

**Comment by BruceWayne233** [comment]
Hi Mr Thomas
wow,I am excited to get your response so quickly.
the iou of ground is around 68，the iou around non_ground is around 91, I want to make the model to handle the ground point perform better. 
This is part of this data.
![image](https://user-images.githubusercontent.com/44676054/147673484-e4f90f17-b774-439e-abe9-dbdaea3dd3de.png)
![image](https://user-images.githubusercontent.com/44676054/147673491-e1a88d85-fe0b-4723-9a12-4c0613d3b72b.png)

Intuitionally, I think using a larger in _ radius and a smaller grid _ subsampling may help this problem ( with enough memory ), but it also needs to be controlled between 50. The parameters I currently use are 
in _ radius = 1.2 
grid _ subsampling = 0.03 
The parameters are not changed. 
How will you handle this problem ? I sincerely want your advice.
By the way,this problem may seem silly，Will all test results be saved in the ply file after the test is completed ? For example, I use area5 as the test data, will all the test results be written back to area5.py ? If so, how do I show this result by cloud compare( showing the label I predicted ) ?
Thank you again so timely reply, wish you all good.

**Comment by HuguesTHOMAS** [comment]
> Intuitionally, I think using a larger in _ radius and a smaller grid _ subsampling may help this problem ( with enough memory ), but it also needs to be controlled between 50. The parameters I currently use are
in _ radius = 1.2
grid _ subsampling = 0.03

Ok so from what I see, it seems that a larger `in_radius` would indeed help with the results, it would help the network catch more features from the trees. In your case, I would try something like 3 meters, and a corresponding `first_subsampling_dl` of 8 cm. You can try to have a bigger ratio between the two values by reducing the `batch_num` to 3 or 4 (I would not advise getting lower than 3 though). You can try out multiple values for these parameters and compare the results with the plot_convergence.py script.
https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/73e444d486cd6cb56122c3dd410e51c734064cfe/plot_convergence.py#L779-L784

> Will all test results be saved in the ply file after the test is completed ? For example, I use area5 as the test data, will all the test results be written back to area5.py ? If so, how do I show this result by cloud compare( showing the label I predicted ) ?

If you choose area5 as the test data, there will be a folder created when training the network at `your_root_folder/results/Log_YYYY-MM-DD_HH-MM-SS`  (corresponding to the date of the training). In this folder you will find the network logs, checkpoints, and the validation results. New _area5.ply_ files will be saved every 50 epochs with the current validation results. You can just load them with CloudCompare they contain the predictions

**Comment by BruceWayne233** [comment]
Hi Mr Thomas
Thank you sincerely for your advice on how to choose parameters.

> If you choose area5 as the test data, there will be a folder created when training the network at your_root_folder/results/Log_YYYY-MM-DD_HH-MM-SS (corresponding to the date of the training). In this folder you will find the network logs, checkpoints, and the validation results. New area5.ply files will be saved every 50 epochs with the current validation results. You can just load them with CloudCompare they contain the predictions

To verify that I understand correctly，when I run test_models.py, every 50 epochs I will get a ply file in the folder Log_YYYY-MM-DD_HH-MM-SS（corrrsponding to the date of the training） ? Do I understand it correctly?

Another question is why the epoch number is 50  ? Where can this parameter be controlled ?

Finally, thank you very much for your help and wish you all the best

**Comment by BruceWayne233** [comment]
Hi Mr Thomas

Antoher quick question,do I need to set the in_radius for test proce?  Will it keep the same parameters as the training procss?
If  I have to manual setting，how could I set it?

Thank you very much for your help and wish you all the best
